virtualizarr/readers/zarr.py

```diff
@@ -0,0 +1,329 @@
+import json
+from pathlib import Path
+from typing import Iterable, Mapping, Optional
+
+import numcodecs
+import numpy as np
+from xarray import Dataset, Index, Variable
+
+from virtualizarr.manifests import ChunkManifest, ManifestArray
+from virtualizarr.readers.common import (
+    VirtualBackend,
+    construct_virtual_dataset,
+    open_loadable_vars_and_indexes,
+    separate_coords,
+)
+from virtualizarr.utils import check_for_collisions
+from virtualizarr.zarr import ZArray
+
+
+class ZarrVirtualBackend(VirtualBackend):
+    @staticmethod
+    def open_virtual_dataset(
+        filepath: str,
+        group: str | None = None,
+        drop_variables: Iterable[str] | None = None,
+        loadable_variables: Iterable[str] | None = None,
+        decode_times: bool | None = None,
+        indexes: Mapping[str, Index] | None = None,
+        reader_options: Optional[dict] = None,
+    ) -> Dataset:
+        """
+        Create a virtual dataset from an existing Zarr store
+        """
+
+        # check that Zarr is V3
+        # 1a
+        import zarr
+        from packaging import version
+
+        if version.parse(zarr.__version__).major < 3:
+            raise ImportError("Zarr V3 is required")
+
+        # check_for_collisions will convert them to an empty list
+        drop_variables, loadable_variables = check_for_collisions(
+            drop_variables,
+            loadable_variables,
+        )
+
+        # can we avoid fsspec here if we are using zarr-python for all the reading?
+
+        # 1b.
+
+        zg = zarr.open_group(filepath, mode="r")
+
+        # 2a. Use zarr-python to list the variables in the store
+        zarr_arrays = [val for val in zg.keys()]
+
+        # 2b. and check that all loadable_variables are present
+        assert set(
+            loadable_variables
+        ).issubset(
+            set(zarr_arrays)
+        ), f"loadable_variables ({loadable_variables}) is not a subset of variables in existing Zarr store. This zarr contains:  {zarr_arrays}"
+
+        # virtual variables are available variables minus drop variables & loadable variables
+        virtual_vars = list(
+            set(zarr_arrays) - set(loadable_variables) - set(drop_variables)
+        )
+
+        virtual_variable_mapping = {}
+        all_array_dims = []
+
+        # 3. For each virtual variable:
+        for var in virtual_vars:
```

Comment:
Everything inside here is for a single array so should also probably be refactored into a function